doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+hello, or it may terminate TLS if it is configured with a certificate for the
+client-requested name (i.e. for external/ingress).
+
+If the proxy elects to terminate TLS, the client's ALPN[[3]](#tls-q3) advertisement is
+used to determine which protocol should be used. If the ALPN advertisement does not
+include a proxy-supported protocol (i.e. `http` or `h2`), then the proxy handles the
+decrypted stream as opaque TCP[4](#tls-q4).
+
+##### Questions
+
+<a name="tls-q1"></a>
+1. What's the minimum version of the TLS protocol we support? I assume we don't handle SSL
+    3.0, etc. due to ClientHello requirement
+<a name="tls-q2"></a>
+2. How do TLS extensions play into this detection? are they conduit-specific or are we
+   reusing existing public extensions?
+<a name="tls-q3"></a>
+3. Can we reasonably expect clients to support ALPN? Do we have to support detecting
+   protocols for TLS requests that don't advertise a protocol?
+<a name="tls-q4"></a>
+4. It's not immediately clear to me how the proxy can "bridge" two TLS
+   connections--propagating ALPN information could be complex: we might want to issue a
+   `ClientHello` to the destination before issuing a ServerHello to the initiatior. This is
+   mostly germane for ingress.
+
+
+#### HTTP
+
+If the connection begins with an HTTP/1 request-line, the connection is handled as HTTP/1.
+
+If the connection begins with a HTTP/2 Prior-knowledge preface, the connection is handled
+as HTTP/2.
+
+
+## TCP Forwarding
+
+If, after protocol detection (and potentially TLS decryption), the connection's protocol
+is not supported by the proxy, the connection is simply forwarded to the original
+destination.
+
+Policy may require that the forwarded connections are TLS'd by the proxy (i.e. for mutual
+authentication).
+
+
+## HTTP Proxying
+
+### HTTP/1.0
+
+If the request is an HTTP/1.0 request that does not include a `Host` header, the request
+is routed to the connection's original destination address. The request is sent on a
+dedicated one-off connection (in keeping with the HTTP/1.0 specification).
+
+If an HTTP/1.0 request is received and _does_ include a `Host` header, the request is
+routed as if it were HTTP/1.1, though the proxy must be sure to honor proper HTTP/1.0
+semantics to the client peer.
+
+### HTTP/1.1
+
+When the proxy receives an HTTP/1.1 request, it is checked to see if the `Connection`
+header includes `Upgrade` in its list of values. If the `Upgrade` header exists, and its
+value is `h2c`, the connection is upgraded to HTTP/2 before continuing. If it has any
+other value, then the request should fail with a 500[1](#http-1).
+
+The proxy honors hop-by-hop headers (i.e. `Connection`, `Authorization`) and it removes
+them from requests before they are routed.
+
+The proxy honors `CONNECT` requests as any other. Once the request is routed to an
+appropriate destination, it is dispatched, and the accepted connection is treated as part
+of the request.
```

Comment:
I don't know what this means. If the server agrees to complete the CONNECT then the remaining traffic on the connection is some arbitrary TCP stream. If the server doesn't agree to complete the CONNECT then the remaining traffic is HTTP.